Feature: Label provider function

codeandcats opened the pull request:
# Summary
- Adds new function `registerLabelTransformer` which accepts a function that accepts a property key and class and returns a user friendly label for that property.
- Add `.vscode` directory to git ignore
- Renames `test/unit/constraints/any.ts` to `test/unit/constraints/any.test.ts` so it gets picked up by jest
- Removes nested `it` calls from `any.test.ts`

# Motivation
Joi's error messages are mostly user friendly except for the fact that they are based on property keys. A consumer can specify user friendly labels for properties via the `Label` decorator which is good. However if most of the time the label you want is simply the property key as title or sentence case, specifying a label decorator on every property isn't very DRY (wet?) and doesn't make for a great developer experience. Having the ability to specify a label transform function once at app startup spares the developer from having to specify labels for every field.

laurence-myers commented on test/unit/core.test.ts:
```diff
@@ -0,0 +1,61 @@
+
+    describe('when a label transform has been supplied', () => {
+        beforeEach(() => {
+            labelTransformer = jest.fn().mockImplementation((propertyKey) => `${propertyKey}`.toUpperCase());
```
❔ Why use `jest.fn()` instead of just a normal function?

laurence-myers commented on src/core.ts:
```diff
@@ -9,6 +9,24 @@ export function registerJoi(customJoi : typeof joi) {
 
+export type LabelTransformer = ((propertyKey: string | Symbol, target: object) => string | null | undefined);
+
+let globalLabelTransformer: LabelTransformer | undefined | null = undefined;
```
:worried: I'm reluctant to add a global variable, we already have one for the Joi instance and I don't much like that. I don't know off-hand a better way to achieve this, though.

Global variables will limit the ability to use tsdv-joi across multiple dependencies within a containing app. If one dependency desires this label transformer, but another dependency is not expecting it and has quirky logic to inspect errors by the label (for some reason), things could break.

laurence-myers replied:
Maybe you could pass it as an option to the validator functions/classes?

laurence-myers commented on .gitignore:
```diff
@@ -4,6 +4,9 @@
 gen
 
+# VSCode
+.vscode
```
👏

laurence-myers commented:
I like this idea. I think using a generated friendly name in error messages is much more useful than the raw property name, and something I wish Joi did natively.

I think it's less useful if your app requires i18n, or if your app's display language is not the same as the language used in the code. (I believe many non-English speaking countries will write code using English identifiers, so an object's property name could be in English, but should be displayed in e.g. German.)

My other concern is that it extends the scope of tsdv-joi beyond just providing decorators for Joi.

codeandcats commented:
I actually think if anything it probably improves the i18n scenario because it allows one to specify a label transform that could perform the i18n lookup based on the classname and property name. The trick would be knowing what locale to use though.

I take your point about the global function and agree specifying a label transform in the validator makes more sense. This would also make i18n easier because you would more likely know the locale at the time of calling validate. Let me make some tweaks.

codeandcats commented:
Note: "Label Transform/Transformer" isn't the best term since it actually doesn't transform existing labels. Rather it provides labels for properties without labels already, thus I'll refer to it as a "Label Provider" from here on.

I had a crack (/hack) at specifying a label provider in the options parameter for `validate`, `validateAsClass` and `validateArrayAsClass` methods of `Validator`, then by getting the `WorkingSchema` for the class, iterating over the property keys, getting each key's schema and checking if a label exists, and if not, updating the key's schema with the label, and finally rebuilding the object schema.

```typescript
function getSchemaWithLabels(clz: AnyClass, labelTransformer: LabelTransformer | undefined) {
    if (!labelTransformer) {
        return getJoiSchema(clz);
    }

    const workingSchema = getMergedWorkingSchemas(clz.prototype);
    for (const key of Object.keys(workingSchema)) {
        let keySchema = workingSchema[key];
        const description = getSchemaDescription(keySchema);
        if (description.label === undefined) {
            const label = labelTransformer(key, clz);
            if (label && label !== key) {
                keySchema = keySchema.label(label);
                workingSchema[key] = keySchema;
            }
        }
    }

    const schemaWithLabels = Joi.object().keys(workingSchema);

    return schemaWithLabels;
}
```

The problem with this approach is it does not update labels of nested properties. Any ideas?

codeandcats commented:
Hey @laurence-myers I might close this if you don't see a benefit / think the implementation should live in user-land or joi itself?

laurence-myers commented:
Sorry for the delay getting back to you.

Good point about the i18n! I hadn't understood the way the transformer function worked.

I'm open to including this in tsdv-joi, if we can iron out the details. Maybe it's not a real problem to have a global label transform. I'll need to have another look over what you've got so far.

Something I mulled over was whether tsdv-joi could expose a "decorator factory" function. So, instead of importing constraints directly, you import them from some module that invokes the decorator factory. There would be a default decorator factory, exposing all the tsdv-joi decorators. If a consuming application wants to apply something within their app, without affecting any dependencies that might also be using tsdv-joi, they can define their own decorator factory, which can (somehow) apply globals like the label transform function or a custom version of Joi.

ben-homepass commented:
Okay, cool. 👍

Interesting idea with the decorator factory, I like it. Just checking I'm on the same page, do you mean like, for example, in axios how you can conveniently [use methods directly](https://github.com/axios/axios#example) or create your [own axios client](https://github.com/axios/axios#creating-an-instance) with specific settings? Make sense. 👍

codeandcats commented:
Okay, cool. 👍

Interesting idea with the decorator factory, I like it. Just checking I'm on the same page, do you mean like, for example, in axios how you can conveniently [use methods directly](https://github.com/axios/axios#example) or [create your own axios client](https://github.com/axios/axios#creating-an-instance) with specific settings? Make sense. 👍

codeandcats commented:
Hey @laurence-myers, I've just force-pushed some commits:
- Rebased this branch off latest master
- Renamed all references to "label transformer" to the more accurate wording of "label provider"
- Added JavaDoc comment to `registerLabelProvider` function

If you're happy to do so, could we please merge this branch in now and deal with the ramifications of global side-effects as part of #41?

codeandcats commented:
Closing this PR in favour of #54.
